### PR TITLE
Introduce container allocation timeout

### PR DESCRIPTION
**Pull request by zuston**

### CODE REVIEW

**Comment by zuston** [comment]
Changelog
1. Introduce checkHealthy method in `Framework.ApplicationMasterAdapter` interface
2. Introduce checking container allocation timeout in checkHealthy method implemented by MLGenericFramework

@oliverhu

**Comment by oliverhu** [comment]
looks good, pls add some tests

File: tony-core/src/test/java/com/linkedin/tony/TestTonyE2E.java [hunk 1]
```diff
@@ -586,6 +586,23 @@ public void testTonyHorovodWithDebugModeShouldPass() throws ParseException, IOEx
...
+    client.init(new String[]{
+            "--src_dir", "tony-core/src/test/resources/scripts",
+            "--executes", "python check_env_and_venv.py",
+            "--hdfs_classpath", libPath,
+            "--shell_env", "ENV_CHECK=ENV_CHECK",
+            "--container_env", Constants.SKIP_HADOOP_PATH + "=true",
+            "--python_venv", "tony-core/src/test/resources/test.zip",
+            "--conf", "tony.worker.instances=10000",
+            "--conf", "tony.worker.memory=2g",
+            "--conf", "tony.container.allocation.timeout=10000"
```
Comment (oliverhu) [inline tony-core/src/test/java/com/linkedin/tony/TestTonyE2E.java:600]: just 1s?

Comment (oliverhu) [reply tony-core/src/test/java/com/linkedin/tony/TestTonyE2E.java:600]: 10s is a waste of time?

Comment (zuston) [reply tony-core/src/test/java/com/linkedin/tony/TestTonyE2E.java:600]: Got it.

File: tony-core/src/main/java/com/linkedin/tony/Framework.java [hunk 2]
```diff
@@ -46,6 +46,11 @@
         /** For AM, it will pre-check tony conf and inject some params. like horovod runtime will inject driver config into it. **/
         boolean validateAndUpdateConfig(Configuration tonyConf);
 
+        /**
+         * For AM, it will check the runtime healthy periodically. If not ok, the job will fast fail.
+         */
+        boolean checkHealthy(Configuration tonyConf);
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/Framework.java:52]: isHealthy()

**Review by oliverhu — APPROVED** [review]
a minor comment

**Pull request merged**